edgelet/edgelet-core/src/module.rs

```diff
@@ -278,21 +281,103 @@ pub trait ModuleRegistry {
     async fn remove(&self, name: &str) -> Result<(), Self::Error>;
 }
 
-#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+#[skip_serializing_none]
+#[derive(Debug, Default, PartialEq, Serialize)]
 pub struct SystemInfo {
-    /// OS Type of the Host. Example of value expected: \"linux\" and \"windows\".
     #[serde(rename = "osType")]
-    pub os_type: String,
-    /// Hardware architecture of the host. Example of value expected: arm32, x86, amd64
+    pub kernel: String,
+    pub kernel_release: String,
+    pub kernel_version: String,
+
+    pub operating_system: Option<String>,
+    pub operating_system_version: Option<String>,
+    pub operating_system_variant: Option<String>,
+    pub operating_system_build: Option<String>,
+
     pub architecture: String,
-    /// iotedge version string
+    pub cpus: usize,
+    pub virtualized: String,
+
+    pub product_name: Option<String>,
+    pub system_vendor: Option<String>,
+
     pub version: String,
+
     pub provisioning: ProvisioningInfo,
-    pub server_version: String,
-    pub kernel_version: String,
-    pub operating_system: String,
-    pub cpus: i32,
-    pub virtualized: String,
+
+    #[serde(default, flatten)]
+    pub additional_properties: BTreeMap<String, String>,
+}
+
+impl SystemInfo {
+    pub fn from_system() -> Result<Self, Error> {
+        let kernel = nix::sys::utsname::uname();
+        let dmi = DmiInfo::default();
+        let os = OsInfo::default();
+
+        let res = Self {
+            kernel: kernel.sysname().to_owned(),
+            kernel_release: kernel.release().to_owned(),
+            kernel_version: kernel.version().to_owned(),
+
+            operating_system: os.id,
+            operating_system_version: os.version_id,
+            operating_system_variant: os.variant_id,
+            operating_system_build: os.build_id,
+
+            architecture: os.arch.to_owned(),
+            cpus: num_cpus::get(),
+            virtualized: match crate::virtualization::is_virtualized_env() {
+                Ok(Some(true)) => "yes",
+                Ok(Some(false)) => "no",
+                _ => "unknown",
+            }
+            .to_owned(),
+
+            product_name: dmi.product,
+            system_vendor: dmi.vendor,
+
+            version: crate::version_with_source_version(),
```

Comment:
Do we need `.to_owned()` here?

Reply:
No, not on main: https://github.com/Azure/iotedge/blob/main/edgelet/edgelet-core/src/lib.rs#L64-L66